runtime/runtime/src/actions.rs

```diff
@@ -624,6 +627,202 @@ pub(crate) fn action_add_key(
+    if result.result.is_err() {
+        return Ok(());
+    }
+
+    let actions = delegate_action.get_actions();
+
+    let new_receipt = Receipt::new_delegate_actions(
+        &action_receipt.signer_id,
+        sender_id,
+        &delegate_action.receiver_id,
+        &actions,
+        &action_receipt.signer_public_key,
+        action_receipt.gas_price,
+    );
+
+    // Note, Relayer prepaid all fees and all things required by actions: attached deposits and attached gas.
+    // If something goes wrong, deposit is refunded to the predecessor, this is sender_id/Sender in DelegateAction.
+    // Gas is refunded to the signer, this is Relayer.
+    // Some contracts refund the deposit. Usually they refund the deposit to the predecessor and this is sender_id/Sender from DelegateAction.
+    // Therefore Relayer should verify DelegateAction before submitting it because it spends the attached deposit.
+
+    let prepaid_send_fees = total_prepaid_send_fees(
+        &apply_state.config.transaction_costs,
+        &action_receipt.actions,
+        apply_state.current_protocol_version,
+    )?;
+    let required_gas = receipt_required_gas(apply_state, &new_receipt)?;
+    // This gas will be burnt by the receiver of the created receipt,
+    result.gas_used = safe_add_gas(result.gas_used, required_gas)?;
+    // This gas was prepaid on Relayer shard. Need to burn it because the receipt is goint to be sent.
+    // gas_used is incremented because otherwise the gas will be refunded. Refund function checks only gas_used.
+    result.gas_used = safe_add_gas(result.gas_used, prepaid_send_fees)?;
+    result.gas_burnt = safe_add_gas(result.gas_burnt, prepaid_send_fees)?;
+    result.new_receipts.push(new_receipt);
+
+    Ok(())
+}
+
+fn receipt_required_gas(apply_state: &ApplyState, receipt: &Receipt) -> Result<Gas, RuntimeError> {
+    Ok(match &receipt.receipt {
+        ReceiptEnum::Action(action_receipt) => {
+            let mut required_gas = safe_add_gas(
+                total_prepaid_exec_fees(
+                    &apply_state.config.transaction_costs,
+                    &action_receipt.actions,
+                    &receipt.receiver_id,
+                    apply_state.current_protocol_version,
+                )?,
+                total_prepaid_gas(&action_receipt.actions)?,
+            )?;
+            required_gas = safe_add_gas(
+                required_gas,
+                apply_state.config.transaction_costs.action_receipt_creation_config.exec_fee(),
+            )?;
+
+            required_gas
+        }
+        ReceiptEnum::Data(_) => 0,
+    })
+}
+
+fn validate_delegate_action_key(
+    state_update: &mut TrieUpdate,
+    apply_state: &ApplyState,
+    delegate_action: &DelegateAction,
+    result: &mut ActionResult,
+) -> Result<(), RuntimeError> {
+    // 'delegate_action.sender_id' account existence must be checked by a caller
+    let mut access_key = match get_access_key(
```

Comment:
line 726 -> 762 is very similar to code in verify_and_charge_transaction method. (except that here we return a little bit different error messages). I wonder if we should somehow combine the two (might be done in separate pr)